CMakePresets.json

```diff
@@ -0,0 +1,65 @@
+{
+  "version": 3,
+  "cmakeMinimumRequired": {
+    "major": 3,
+    "minor": 21,
+    "patch": 0
+  },
+  "configurePresets": [
```

Comment:
(Future cleanup? Possibly worth splitting some of this into a more focused PR)

We could union the builds in https://github.com/google/iree/blob/main/build_tools/buildkite/cmake/build_configurations.yml with this list of presets (define presets for all configurations in that file, add a test step for each preset, remove any configurations that aren't important enough to be in both?)

As part of that we could remove https://github.com/google/iree/blob/main/build_tools/cmake/build_runtime.sh, https://github.com/google/iree/blob/main/build_tools/cmake/build_tracing.sh, and https://github.com/google/iree/blob/main/build_tools/cmake/build_runtime_small.sh in favor of forwarding `--preset` through to a single script.

* presets should be tested
* we should aim to have one preferred way of configuring the build system

Reply:
Aiming is fine. Getting there in a few steps is preferred. I have very little mileage on this presets mechanism and we need to kick the tires a bit. Would prefer to land here since how it works actually influences the nature of our CMake options in a way that is hard to see otherwise.

Reply:
SGTM. Those scripts are really just a way to set flags like

```bash
"${CMAKE_BIN?}" -G Ninja .. \
  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
  -DIREE_BUILD_COMPILER=OFF
```

```bash
"${CMAKE_BIN?}" -G Ninja .. \
  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
  -DIREE_ENABLE_RUNTIME_TRACING=ON \
  -DIREE_BUILD_COMPILER=OFF
```

on our CI. Swapping the scripts for `--preset` would be pretty straightforward.

Some of this may also intersect with Geoffrey's automation/CI refactoring.